# knitr document van Steensel lab

# Can we detect hopping by tagmentation?
## Christ Leemans, 01-29-2020

## Introduction.

Local hopping of transposons can be used to introduce random disruptions of
in the vicinity of a region of interest. Using cre-recombinase Omar previously
attempted to systematically rearrange random parts of the genome including
LAD-borders. This was previously done by iPCR, which was laborious and error-prone.
Now Lise is giving this project another go using tagmentation.

Another project that was recently started by Mathias involves random transposon
hopping in the hope of disrupting regulatory regions for genes of interest.

To analyze the sequencing data from these experiments I designed a sequencing
pipeline. In both these experiments there is a carrier sequence integrated which
contains the sleeping beauty (SB). This carrier sequence I termed "home".

This carrier can be another transposon (e.g. piggybac) or it could be integrated
using a CRISPR construct. The piggibac (PB) is random, but it is very predictable
which sequence gets build in. For CRISPR at least one site is predictable and
tested by PCRs around that site. But there can be ends of the construct that get
lost or ends of the genome that get resected and lost.

The pipeline should be able to take the location of this home from a tab
delimited file, sanger sequences around the intergration or from another
tagmentation mapping using primers on the home sequence reaching into the genome.

For now I implemented the last input option for the home, the others should
soon follow.

In this first experiment Mathias added different concentrations of the
transposon plasmid to optimize the local hopping.
This was done in a clonal line with PB integrations containing a SB.
He used 5ug, 1ug and a control without plasmid. What we expect to see is when
there is hopping, tagmentation on the SB shows relatively more reads aligning
to parts of the genome outside of the PB intergrations.

Since hopping should also be local events, these reads should also be more
closeby the initial PB sites than reads outside PB for the control.


## Library loading and functions

```{r}
library(data.table)
library(rtracklayer)
library(ggplot2)

read_pairs <- function(row){
    col_vec = c('seqnames', 'start', 'end', 'read_id', 'mapq', 'strand',
                'seqnames_o', 'start_o', 'end_o', 'overlap', 'p_adj',
                'strand_o')
    fwd_dt = fread(paste0('/DATA/usr/c.leemans/projects/tagmap_hopping/',
                          'results/cl20200122_hopping_mathias/mm10/hopping/',
                          'overlap/', row$Forward, '.txt'), sep='\t',
                   col.names=col_vec)
    rev_dt = fread(paste0('/DATA/usr/c.leemans/projects/tagmap_hopping/',
                          'results/cl20200122_hopping_mathias/mm10/hopping/',
                          'overlap/', row$Reverse, '.txt'), sep='\t',
                   col.names=col_vec)
    fwd_dt[,side:='Forward']
    rev_dt[,side:='Reverse']
    return(rbind(fwd_dt, rev_dt))
}

read_insertions <- function(row){
    fread(paste0('/DATA/usr/c.leemans/projects/tagmap_hopping/',
                 'results/cl20200122_hopping_mathias/mm10/home/',
                 'insertions/', row$home_location, '.txt'),
                 sep='\t')
}


get_distance <- function(dt, insertion_gr){
    dt_gr = makeGRangesFromDataFrame(dt)
    n = distanceToNearest(dt_gr, insertion_gr[insertion_gr$ID=='m5ug'])
    result = copy(dt)
    result[queryHits(n),distance:=data.frame(n)$distance]
    return(result)
}

get_median <- function(distance){
    d = ifelse(is.na(distance), Inf, distance)
    return(median(d))
}

```

## loading data


```{r}
## this file contains information on pairing of fwd and reverse and which
## tagmentation of the PB corresponds to the experiment.
pairing_dt = fread(paste0('/DATA/usr/c.leemans/projects/tagmap_hopping/',
                          'config/cl20200121_hopping_mix_pairing.tsv'),
                   sep='\t')[home_sequence=='pJOYC037',]

## the pipeline produces a table with for every read-pair, it's location
## and information whether it overlaps with either a putative integration site
## of the original PB (home), or any other region in the tagmentation experiment
## of the PB.

read_pairs = pairing_dt[,read_pairs(.SD), by='ID']


## this is a file with putative integrations of the PB home and a score
## for how likely these integrations are real.
insertions = pairing_dt[,read_insertions(.SD), by='ID']

insertion_gr = makeGRangesFromDataFrame(insertions[,-c('start', 'end')],
                                        keep.extra.columns=T,
                                        start.field='start_gap',
                                        end.field='end_gap')
```


## reads of potential hopping events.

So first let's look at how much of the reads in the SB tagmap are outside the
PB sites.


```{r}
overlap_count = read_pairs[,data.table(table(overlap)),by=c('ID', 'side')]

overlap_count[, N[overlap=='.']/sum(N[grep('insert_', overlap)]),
              by=c('ID', 'side')]

```

**conclusion:**
Both forward and reverse tagmentation experiments show slightly more reads
found outside of the transposons for the 1ug and 5ug experiments.
However there seems to be very little reads even in the 5ug not originating
from putative hopping events.



## Distance to the PB home sites

Now let's look at the distance of the reads we found to the sites of integration.
We expect for the samples with activated transposase to find more reads closer
to the original PB sites, since that is where we expect the hopped SB to go
preferentially.

```{r}
distance_dt = read_pairs[overlap=='.', get_distance(.SD, insertion_gr),by='ID']

distance_dt[, length(which(!is.na(distance))) / nrow(.SD), by=c('ID', 'side')]


## median distance with NA as Inf (NA are reads coming from chromosomes without PB)
distance_dt[, get_median(distance), by=c('ID', 'side')]



## median distance without NA
distance_dt[, median(distance, na.rm=T), by=c('ID', 'side')]


## ratio of reads with distance < 1MB
distance_dt[, length(which(distance<10^6))/length(distance), by=c('ID', 'side')]

```

**conclusion:**

There is a slight tendency for Forward reads outside of PB from the 5ug sample
to be on the same chromosome and closer to the PB sites.
I expected Reverse reads to be closer to the PB sites, because there seems to
be less noise in those reads because these have less
reads outside of PB in the control group. However this is not the case.

In general most of the reads not linked to a PB integration are very distant,
so might not come from an actual SB hopping, or a very distant SB hopping.


## some extra info

Same calculations, but then Forward and Reverse side grouped together

```{r}

distance_dt[, length(which(!is.na(distance))) / nrow(.SD), by='ID']


## median distance with NA as Inf (NA are reads coming from chromosomes without PB)
distance_dt[, get_median(distance), by='ID']



## median distance without NA
distance_dt[, median(distance, na.rm=T), by='ID']


## ratio of reads with distance < 1MB
distance_dt[, length(which(distance<10^6))/length(distance), by='ID']

```

## Density plots

```{r}

ggplot(distance_dt, aes(x=log10(distance), color=paste(ID, side))) +
    geom_density(adjust=.5) +
    theme_bw()

ggplot(distance_dt, aes(x=log10(distance), color=ID)) +
    geom_density(adjust=.5) +
    theme_bw()

```
